doc/specs/#16599 - Quick Fix.md

```diff
@@ -0,0 +1,58 @@
+- WinGet Command Not Found OSC Sequence:
+	- `OSC 9001; CmdNotFound; <missingCmd>` will be used by the attached CLI app to notify the terminal that a command `missingCmd` wasn't found. The TerminalControl layer will use WinGet's API to search for any packages that would remediate this error and construct a list of suggested packages to be installed.
+`OSC 9001` can be expanded on to add more suggestions via VT sequences from the underlying CLI app; for more on this, read the [Future considerations section](#future-considerations).
+
+Upon receiving an `OSC 9001; CmdNotFound`, the `missingCmd` parameter will be stored in the text buffer's `ROW`. Specifically, as of PR \#16937, `ROW` will contain an optional `ScrollbarData` struct to hold any metadata (i.e. scrollbar marks). We can stuff a `std::wstring missingCmd` into that metadata and populate it if `OSC 9001; CmdNotFound` was received.
+
+`ControlCore` will be notified upon receiving an `OSC 9001; CmdNotFound`. At this point, we'll know that a Quick Fix menu must be shown at the next prompt, and we can use the WinGet API to convert the given `missingCmd` into a list of suggestions (in this case, of the form `winget install...`). Since we're in the TerminalControl layer, the suggestions will not need to be converted into SendInput actions, as we already have access to the `RawWriteString()` api. The menu will simply use `RawWriteString()` to inject the string into the prompt.
+
+A corner case that VS Code handles is that if there is already text in the prompt area, it sends a `^C`, then injects the suggestion, then executes it. Terminal will have to operate the same way and, thus, must also be able to detect if there is already text in the prompt area.
+
+An additional corner case is handling a buffer resize and reflow. Since the `missingCmd` will be stuffed as metadata in the text buffer's `ROW`, we should be able to generate the WinGet suggestions again fairly easily, if needed. Ideally, the generated suggestions can be cached and the quick fix menu is simply redrawn in the new reflowed position.
+
+The WinGet API to convert the given `missingCmd` into a list of suggestions may be relatively slow as we have to query the WinGet database. Performing the search when the quick fix menu is open would not be ideal as the user may have to wait for WinGet to finish searching. Instead, the conversion should happen around the time that the prompt is introduced (whether from output or a reflow scenario).
+
+## UI/UX Design
+A Quick Fix icon will be displayed immediately left of the prompt. If the icon fits in the left-padding region (a `padding` value where the left side is 16 or greater), it will be drawn there. Otherwise, a thin accent-colored vertical bar will be displayed overlapping the text region; when the icon is hovered over, the icon will appear more clear by becoming more opaque and enlarging to overlap with the text region.
```

Comment:
accent color is an interesting choice - why that over like, the foreground color? Or yellow from the scheme's color table?

Reply:
Personally, I feel like accent color is a good way to signify that the UI is personalized to the user from _by the system_ (vs color scheme colors being a part of the standard theming). Accent colors just pop more, imo.

Reply:
@zadjii-msft here's the UI I have from my branch.

![quickFix demo](https://github.com/microsoft/terminal/assets/11050425/ceaae0aa-f04e-402f-947d-0a94e1e42a88)

![quickFix demo (collapsed)](https://github.com/microsoft/terminal/assets/11050425/5886ece1-e42a-4632-ac3a-6d74f51daa71)

Reply:
oh my god it's so cool

Reply:
> @zadjii-msft here's the UI I have from my branch.

Looks slick! One question though - is the button supposed to show up on the prompt with the error in it? The GIF shows the button showing up on the _next_ empty prompt line, which IMO is a little unintuitive

Reply:
Yup! Intentional to align with VS Code's design. I think it's so that if you had _a lot_ of output, you don't have to scroll all the way back up to get to the quick fix menu.

Reply:
That makes sense! Should the suggestion get input into that empty line then? Like in the image below. (Not sure if that's just a demo bug, just making sure I'm not missing it if it's in the spec!)

<img width="402" alt="image" src="https://github.com/microsoft/terminal/assets/26824113/d656a308-ee40-4ca5-91fb-abaf31d0592a">

Reply:
eh, it's probably because of the secret ^C we insert before the command to make sure it doesn't conflict with input the user has already typed

Reply:
> eh, it's probably because of the secret ^C we insert before the command to make sure it doesn't conflict with input the user has already typed

Yup, exactly that. We're always sending over the ^C for now, which is why we end up getting that blank line. But hopefully sometime in the future (perhaps with shell integration) we won't need to do that always.